text/2020-11-09-chaos-mesh-workflow.md

```diff
@@ -0,0 +1,293 @@
+# Workflow
+
+## Summary
+
+A built-in workflow engine with frontend support is needed for Chaos Mesh.
+
+## Motivation
+
+Both running chaos experiments parallelly or sequentially are important for
+simulating production errors. Chaos Mesh is shipped with a feature-rich chaos
+toolbox. Composing them together can make it much more powerful. With a workflow
+engine, an experiment can turn from single chaos into a series of chaos with
+some external logic (like checking health).
+
+Previously, the `duration` and `scheduler` fields have enabled a very
+fundamental "workflow" feature, so after implementing the workflow engine, these
+fields should be deprecated.
+
+## Detailed design
+
+The specification for a `Workflow` custom resource can be concluded in the
+following example:
+
+```yaml
+spec:
+  entry: serial-chaos
+  templates:
+  - name: networkchaos
+    type: NetworkChaos
+    duration: "30m"
+    selector:
+      labelSelectors:
+        "component": "tikv"
+    delay:
+      latency: "90ms"
+      correlation: "25"
+      jitter: "90ms"
+  - name: iochaos
+    type: IOChaos
+    duration: "10m"
+    selector:
+      labelSelectors:
+        app: "etcd"
+    volumePath: /var/run/etcd
+    path: /var/run/etcd/**/*
+    errno: 5
+    percent: 50
+  - name: parallel-chaos
+    type: Parallel
+    tasks:
+    - ref: networkchaos
+    - ref: iochaos
+  - name: check
+    type: Task
+    task:
+      image: 'alpine:3.6'
+      command: ['/app']
+    branch:
+    - when: "stdout == 'continue'"
+      ref: serial-chaos
+    - when: "stdout == 'continue'"
+      ref: serial-chaos
+  - name: serial-chaos
+    type: Serial
+    deadline: "30m"
+    tasks:
+    - ref: networkchaos
+    - ref: iochaos
+    - ref: parallel-chaos
+    - type: Suspend
+      duration: "30m"
+    - ref: check
+```
+
+### `templates` field
+
+`templates` field describes all the templates. One of them is the entry of the
+workflow (which is specified by the `entry` field). The `templates` is a slice
+of template, each of them is a `*Chaos`, `Parallel`, `Serial` or `Task`. They
+are distinguished according to the `type` field.
```

Comment:
The terminologies here, TBH, make me confused. `*Chaos` and `Task` is the real task execution spec, and `Parallel ` and `Serial` describe the execution steps, so we would make the spec like the following example, as you offered:

```
...
    spec:
    - ref: networkchaos
    - ref: iochaos
    - ref: parallel-chaos
    - type: Suspend
      duration: "30m"
    - ref: check
...
```

I think it's not intuitive, the new coming users would be troubled on understanding the meaning of the workflow spec.

From argo, it has `step` and `template`, step describe the execution flow, and template is the execution unit, and it use concurrency policy to control the parallelism.

I'd like to share one more well-designed workflow language, see OpenStack mistral workflow engine [language v2](https://docs.openstack.org/mistral/latest/user/wf_lang_v2.html), it also has something like `step` and `template`, in its scenario, it called them as `Task` and `Action`, what's more, if we want to run a task actions on parallel mode, we can define the concurrency value (default to be 1, which means serial run), make it to sth like 2 or N, and it will run the actions on parallel mode.

In a word, I suggest separate these two kind of semantics into their own field, rather than use `type` field, it would make our workflow spec not intuitive.

Reply:
🍻 Thanks for your suggestion! 

In `Argo`, both `steps` and `dag` control the execution flow (however, we don't plan to support `dag` in Chaos Mesh), and they are all one kind of template. And other kinds of templates are execution unit. As I know they didn't seperate the execution flow out of `template`.

The only difference between Argo and this Chaos Mesh design is that: Argo distinguishes different kinds of templates through the JSON field, e.g. the if this template has a `steps` field, it's a `steps` template, if it has a `container` field, it's a `container` template, but in Chaos Mesh Workflow, they are distinguished through `type`.

This difference comes from two different ways to convert a tagged enum into JSON. The way described in this RFC is called "internally tagged" and the way Argo used is called "externally tagged". After reconsideration, I agree with you that the "externally tagged" could be more intuitive in this senario.

I will look into "Mistral Workflow Language (v2)" to see what can we learn from its design. Thanks again for providing this learning material 🤝